src/librustc/traits/error_reporting.rs

```diff
@@ -410,6 +409,37 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
             flags.push(("crate_local".to_owned(), None));
         }
 
+        // Allow targetting all integers using `{integral}`, even if the exact type was resolved
+        if self_ty.is_integral() {
+            flags.push(("_Self".to_owned(), Some("{integral}".to_owned())));
+        }
+
+        if let ty::Array(aty, len) = self_ty.sty {
+            flags.push(("_Self".to_owned(), Some("[]".to_owned())));
+            flags.push(("_Self".to_owned(), Some(format!("[{}]", aty))));
+            if let Some(def) = aty.ty_adt_def() {
+                // We also want to be able to select the array's type's original
+                // signature with no type arguments resolved
+                flags.push((
+                    "_Self".to_owned(),
+                    Some(format!("[{}]", self.tcx.type_of(def.did).to_string())),
+                ));
+                if let Some(len) = len.val.try_to_scalar().and_then(|scalar| {
+                    scalar.to_i64().ok()
```

Comment:
Hi @oli-obk!

Could you take a look at how I'm using `to_i64` here? In travis I'm getting the following test error, and looking at [the interface of `Scalar`](https://github.com/rust-lang/rust/blob/master/src/librustc/mir/interpret/value.rs#L114) I don't see a safe way of figuring wether the value is `1` without tripping the assert on some platforms. I need to check wether `len.val` is `1` (at the very least) or the actual value (more flexibility, but not needed now).

```
[00:46:27] thread 'main' panicked at 'assertion failed: `(left == right)`
[00:46:27]   left: `8`,
[00:46:27]  right: `4`', librustc/mir/interpret/value.rs:242:17
[00:46:27] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[00:46:27] 
[00:46:27] error: internal compiler error: unexpected panic
```

Reply:
You can use [`scalar.to_usize(self.tcx).ok()`](https://github.com/rust-lang/rust/blob/master/src/librustc/mir/interpret/value.rs#L312) which actually takes the target platform's `usize` size and should thus work on all platforms

Reply:
Thanks! I didn't realize that `TyCtxt` implemented `HasDataLayout`.